### PR TITLE
feat: implement multilinear sumcheck data structures

### PR DESCRIPTION
**Pull request by ashjeong**
# Description

- Adds LinearCombination data structure
- Adds VerifyingKey/ProvingKey
- Tests data structures

Note: logic regarding num_variables and related checks is not finalized

Referenced [arkworks' sumcheck code](https://github.com/arkworks-rs/sumcheck/blob/master/src/ml_sumcheck/data_structures.rs)

### CODE REVIEW

**Comment by ashjeong** [comment]
Other notes: 
- `try_emplace` to replace `Contains` is not viable in this 
- `shared_ptrs` used to keep scope n allow for identical ptrs with `.get()`

**Comment by chokobole** [comment]
Please add unittests!

File: tachyon/crypto/sumcheck/multilinear/proof_keys.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+namespace tachyon {
+namespace math {
+
+struct Term;
+
+}  // namespace math
+
+namespace crypto::sumcheck {
+
+struct TACHYON_EXPORT VerifyingKey {
```
[the remark this answers is not in the document]
Comment (ashjeong) [reply tachyon/crypto/sumcheck/multilinear/proof_keys.h:19]: Will do~

File: tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+TEST(LinearCombination, Constructors) {
+  LinearCombination<Poly> a;
+  crypto::sumcheck::VerifyingKey expected_a{0, 0};
+  EXPECT_EQ(a.VerifyingKey(), expected_a);
+  LinearCombination<Poly> b(kMaxDegree);
+  crypto::sumcheck::VerifyingKey expected_b{0, kMaxDegree};
+  EXPECT_EQ(b.VerifyingKey(), expected_b);
+  LinearCombination<Poly> c = b;
+  EXPECT_EQ(c.VerifyingKey(), expected_b);
+}
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc:36]: I think it's too minor to test

File: tachyon/crypto/sumcheck/multilinear/proof_keys_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+namespace tachyon::base {
+
+TEST(VerifyingKey, Copyable) {
+  crypto::sumcheck::VerifyingKey expected =
+      crypto::sumcheck::VerifyingKey::Random();
+
+  std::vector<uint8_t> vec;
+  vec.resize(base::EstimateSize(expected));
+  base::Buffer write_buf(vec.data(), vec.size());
+  ASSERT_TRUE(write_buf.Write(expected));
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/multilinear/proof_keys_unittest.cc:16]: Please apply the recent changed style! See https://github.com/kroma-network/tachyon/pull/300/commits/89980eb533999791d1ea5a00d3620e734ba523a7

File: tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  // `evals` takes in 5 random Polys and an additional 5 of the same Random
+  // Poly `test_multiple` to test logic of inputting the same random Poly
+  std::vector<eval_ptr> evals_1;
+  evals_1.reserve(10);
+  eval_ptr test_multiple = std::make_shared<Poly>(Poly::Random(kMaxDegree));
+  for (size_t i = 0; i < 5; ++i) {
+    evals_1.push_back(std::make_shared<Poly>(Poly::Random(kMaxDegree)));
+    evals_1.push_back(test_multiple);
+  }
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc:52]:
```suggestion
  std::vector<eval_ptr> evals_1 = base::CreateVector(10, [test_multiple](size_t i) {
    if (i % 2 != 0) return std::make_shared<Poly>(Poly::Random(kMaxDegree));
    return test_multiple;
  });
```

File: tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      base::CreateVector(5, std::make_shared<Poly>(Poly::Random(kMaxDegree)));
+  linear_combination.AddTerm(coefficient_2, evals_2);
+  expected_evaluation +=
+      coefficient_2 *
+      std::accumulate(evals_2.begin(), evals_2.end(), F::One(),
+                      [&evaluation_point](F& acc, const eval_ptr eval) {
+                        return acc *= eval->Evaluate(evaluation_point);
+                      });
+  CHECK_EQ(linear_combination.SumOfProducts(evaluation_point),
+           expected_evaluation);
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc:83]:
```suggestion
  EXPECT_EQ(linear_combination.SumOfProducts(evaluation_point),
           expected_evaluation);
```

File: tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  F expected_evaluation =
+      coefficient_1 *
+      std::accumulate(evals_1.begin(), evals_1.end(), F::One(),
+                      [&evaluation_point](F& acc, const eval_ptr eval) {
+                        return acc *= eval->Evaluate(evaluation_point);
+                      });
+
+  // Test validity of `AddTerm()` and `SumOfProducts()` on 1 term
+  CHECK_EQ(linear_combination.SumOfProducts(evaluation_point),
+           expected_evaluation);
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/multilinear/linear_combination_unittest.cc:69]:
```suggestion
  EXPECT_EQ(linear_combination.SumOfProducts(evaluation_point),
           expected_evaluation);
```

File: tachyon/crypto/sumcheck/multilinear/BUILD.bazel [hunk 7]
```diff
@@ -24,3 +24,18 @@ tachyon_cc_library(
         "//tachyon/math/polynomials/multivariate:multilinear_extension",
     ],
 )
+
+tachyon_cc_unittest(
+    name = "data_structures_unittest",
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/multilinear/BUILD.bazel:29]:
```suggestion
    name = "multilinear_unittests",
```

File: tachyon/crypto/sumcheck/multilinear/proof_keys_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,26 @@
+#include "tachyon/crypto/sumcheck/multilinear/proof_keys.h"
+
+#include <vector>
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/multilinear/proof_keys_unittest.cc:3]: If you apply the change below, you can remove this.

File: tachyon/crypto/sumcheck/linear_combination_term.h [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    return std::accumulate(
+        indexes.begin(), indexes.end(), F::One(),
+        [&flattened_ml_evaluations, &point](F& acc, const size_t index) {
+          return acc *= flattened_ml_evaluations[index]->Evaluate(point);
+        });
+  }
+};
+
+}  // namespace math
+namespace base {
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/linear_combination_term.h:61]: Please put blank between namespace

File: tachyon/crypto/sumcheck/sumcheck_proving_key.h [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+  bool operator==(const ProvingKey& other) const {
+    return verifying_key == other.verifying_key && terms == other.terms &&
+           flattened_ml_evaluations == flattened_ml_evaluations;
+  }
+  bool operator!=(const ProvingKey& other) const { return !operator==(other); }
+};
+
+}  // namespace crypto
+namespace base {
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/sumcheck_proving_key.h:38]: Please put blank between namespaces.

**Comment by chokobole** [comment]
Please remove the scope from https://github.com/kroma-network/tachyon/pull/301/commits/2104cb855af2e04c1110b326c7526f211f4040ed since this contains math as well.

File: tachyon/crypto/sumcheck/linear_combination_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    math::MultilinearDenseEvaluations<F, kMaxDegree>>;
+using ProductList = math::LinearCombination<Poly>;
+using Point = Poly::Point;
+using eval_ptr = std::shared_ptr<Poly>;
+
+class LinearCombinationTest : public math::FiniteFieldTest<F> {};
+
+}  // namespace
+
+// Tests `AddTerm()` and `SumOfProducts()`
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/linear_combination_unittest.cc:31]:
```suggestion
// Tests |AddTerm()| and |SumOfProducts()|
```

File: tachyon/crypto/sumcheck/linear_combination_unittest.cc [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+TEST(LinearCombinationTest, AddTermAndSumOfProducts) {
+  F::Init();
+  ProductList product_list(kMaxDegree);
+  F coefficient_1 = F::Random();
+
+  // `evals` takes in 5 random Polys and an additional 5 of the same Random
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/linear_combination_unittest.cc:37]:
```suggestion
  // |evals| takes in 5 random Polys and an additional 5 of the same Random
```

File: tachyon/crypto/sumcheck/linear_combination_unittest.cc [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  // Poly `test_multiple` to test logic of inputting the same random Poly
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/linear_combination_unittest.cc:38]:
```suggestion
  // Poly |test_multiple| to test logic of inputting the same random Poly
```

File: tachyon/crypto/sumcheck/linear_combination_unittest.cc [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+  // Set up the expected evaluation result for this 1 term
+  F expected_evaluation =
+      coefficient_1 *
+      std::accumulate(evals_1.begin(), evals_1.end(), F::One(),
+                      [&evaluation_point](F& acc, const eval_ptr eval) {
+                        return acc *= eval->Evaluate(evaluation_point);
+                      });
+
+  // Test validity of `AddTerm()` and `SumOfProducts()` on 1 term
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/linear_combination_unittest.cc:59]:
```suggestion
  // Test validity of |AddTerm()| and |SumOfProducts()| on 1 term
```

File: tachyon/crypto/sumcheck/linear_combination_unittest.cc [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  EXPECT_EQ(product_list.SumOfProducts(evaluation_point), expected_evaluation);
+
+  // Test `AddTerm()` and `SumOfProducts` on 2 Terms:
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/linear_combination_unittest.cc:62]:
```suggestion
  // Test |AddTerm()| and |SumOfProducts| on 2 terms:
```

File: tachyon/math/polynomials/multivariate/multilinear_dense_evaluations.h [hunk 16]
```diff
@@ -144,8 +146,8 @@ class MultilinearDenseEvaluations {
 
   // Evaluate polynomial at |point|. It uses |FixVariables()| internally. The
   // |point| is a vector in {0, 1}ᵏ in little-endian form. If the size of
-  // |point| is less than the degree of the polynomial, the remaining components
-  // of |point| are assumed to be zeros.
+  // |point| is less than the degree of the polynomial, the remaining
+  // components of |point| are assumed to be zeros.
```
Comment (chokobole) [inline tachyon/math/polynomials/multivariate/multilinear_dense_evaluations.h:150]: Why did you change this?

File: tachyon/math/polynomials/multivariate/multilinear_dense_evaluations.h [hunk 17]
```diff
@@ -169,8 +171,8 @@ class MultilinearDenseEvaluations {
   friend class internal::MultilinearExtensionOp<
       MultilinearDenseEvaluations<F, MaxDegree>>;
 
-  // NOTE(chokobole): This creates a polynomial that contains |F::Zero()| up to
-  // |degree| + 1.
+  // NOTE(chokobole): This creates a polynomial that contains |F::Zero()| up
+  // to |degree| + 1.
```
Comment (chokobole) [inline tachyon/math/polynomials/multivariate/multilinear_dense_evaluations.h:175]: Why did you change this?

File: tachyon/crypto/sumcheck/sumcheck_proving_key_unittest.cc [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "gtest/gtest.h"
+
+#include "tachyon/base/buffer/vector_buffer.h"
+#include "tachyon/crypto/sumcheck/linear_combination.h"
+#include "tachyon/math/finite_fields/test/gf7.h"
+#include "tachyon/math/polynomials/multivariate/multilinear_dense_evaluations.h"
+#include "tachyon/math/polynomials/multivariate/multilinear_extension.h"
+
+namespace tachyon::base {
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/sumcheck_proving_key_unittest.cc:13]:
```suggestion
namespace tachyon::crypto {
```

Comment (ashjeong) [reply tachyon/crypto/sumcheck/sumcheck_proving_key_unittest.cc:13]: is testing copyable supposed to be in `base` or the object's native namespace?

File: tachyon/crypto/sumcheck/BUILD.bazel [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+package(default_visibility = ["//visibility:public"])
+
+tachyon_cc_library(
+    name = "sumcheck_proving_key",
+    hdrs = ["sumcheck_proving_key.h"],
+    deps = [
+        ":sumcheck_verifying_key",
+        "//tachyon/base:random",
+        "//tachyon/base/buffer:copyable",
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/BUILD.bazel:11]: It's up to you, though. You can remove this dependencies because `sumcheck_verifying_key` depends on them

File: tachyon/crypto/sumcheck/sumcheck_proving_key.h [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace tachyon {
+namespace crypto {
+
+template <typename MLE>
+struct ProvingKey {
+  VerifyingKey verifying_key;
+  std::vector<math::LinearCombinationTerm<typename MLE::Field>> terms;
+  std::vector<MLE> flattened_ml_evaluations;
+
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/sumcheck_proving_key.h:20]: Are you going to implement `Random()`?

File: tachyon/math/polynomials/combination/BUILD.bazel [hunk 21]
```diff
@@ -0,0 +1,25 @@
+load("//bazel:tachyon_cc.bzl", "tachyon_cc_library")
+
+package(default_visibility = ["//visibility:public"])
+
+tachyon_cc_library(
+    name = "linear_combination",
+    hdrs = ["linear_combination.h"],
+    deps = [
+        ":linear_combination_term",
+        "//tachyon/base:parallelize",
```
Comment (chokobole) [inline tachyon/math/polynomials/combination/BUILD.bazel:10]: You can safely remove this dependency, though.

File: tachyon/math/polynomials/combination/linear_combination_term.h [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+
+template <typename F>
+struct LinearCombinationTerm {
+  F coefficient;
+  std::vector<size_t> indexes;
+
+  template <typename MLE, typename Container>
+  F Evaluate(
+      const Container& point,
+      const std::vector<std::shared_ptr<MLE>>& flattened_ml_evaluations) const {
```
Comment (chokobole) [inline tachyon/math/polynomials/combination/linear_combination_term.h:23]: Could you match the order with the one in `EvaluateSerial` for readability?

File: tachyon/math/polynomials/combination/linear_combination_term.h [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+  bool operator!=(const LinearCombinationTerm& other) const {
+    return !operator==(other);
+  }
+
+ private:
+  template <typename MLE, typename Container>
+  static F EvaluateSerial(
+      absl::Span<const size_t> indexes,
+      const std::vector<std::shared_ptr<MLE>>& flattened_ml_evaluations,
+      const Container& point) {
```
Comment (chokobole) [inline tachyon/math/polynomials/combination/linear_combination_term.h:51]: In my opinion, it's good to move `indexes` parameter to the end for readability considering the function above.

File: tachyon/math/polynomials/combination/linear_combination.h [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+#include "tachyon/base/parallelize.h"
+#include "tachyon/crypto/sumcheck/sumcheck_proving_key.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifying_key.h"
+#include "tachyon/math/polynomials/combination/linear_combination_term.h"
+
+namespace tachyon {
+namespace math {
+
+template <typename MLE>
+class LinearCombination {
```
Comment (chokobole) [inline tachyon/math/polynomials/combination/linear_combination.h:23]: When seeing the template typename it would be good to place it under `multivariate`! What do you think?

File: tachyon/math/polynomials/combination/linear_combination.h [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  crypto::ProvingKey<MLE> BuildProvingKey() const {
+    std::vector<MLE> flattened_ml_evaluations;
+    flattened_ml_evaluations =
+        base::Map(flattened_ml_evaluations_,
+                  [](const std::shared_ptr<MLE> ptr) { return *ptr.get(); });
+    return {
+        {max_evaluations_, num_variables_}, terms_, flattened_ml_evaluations};
+  }
```
Comment (chokobole) [inline tachyon/math/polynomials/combination/linear_combination.h:48]: Please do not use `crypto` things here!

File: tachyon/crypto/sumcheck/sumcheck_proving_key.h [hunk 26]
```diff
@@ -0,0 +1,63 @@
+#ifndef TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVING_KEY_H_
+#define TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVING_KEY_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/buffer/copyable.h"
+#include "tachyon/base/random.h"
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/sumcheck_proving_key.h:8]: You can safely remove this

File: tachyon/crypto/sumcheck/BUILD.bazel [hunk 27]
```diff
@@ -21,3 +21,28 @@ tachyon_cc_library(
...
+tachyon_cc_unittest(
+    name = "sumcheck_verifying_key_unittest",
+    srcs = [
+        "sumcheck_verifying_key_unittest.cc",
+    ],
+    deps = [
+        ":sumcheck_verifying_key",
+        "//tachyon/base/buffer:vector_buffer",
+    ],
+)
```
Comment (chokobole) [inline tachyon/crypto/sumcheck/BUILD.bazel:48]: Add single unittest called `sumcheck_unittests`!

File: tachyon/math/polynomials/combination/BUILD.bazel [hunk 28]
```diff
@@ -23,3 +23,18 @@ tachyon_cc_library(
         "//tachyon/base/buffer:copyable",
     ],
 )
+
+tachyon_cc_unittest(
+    name = "linear_combination_unittest",
+    srcs = [
+        "linear_combination_unittest.cc",
+    ],
```
Comment (chokobole) [inline tachyon/math/polynomials/combination/BUILD.bazel:31]:
```suggestion
    name = "linear_combination_unittests",
    srcs = ["linear_combination_unittest.cc"],
```

**Review by chokobole — APPROVED** [review]
LGTM! Congratue on the first PR :wave:

File: tachyon/math/polynomials/multivariate/multilinear_extension.h [hunk 29]
```diff
@@ -166,7 +168,33 @@ H AbslHashValue(H h, const MultilinearExtension<Evaluations>& mle) {
...
 }
 
-}  // namespace tachyon::math
+}  // namespace math
+namespace base {
+
+template <typename Evaluations>
+class Copyable<math::MultilinearExtension<Evaluations>> {
+ public:
+  static bool WriteTo(const math::MultilinearExtension<Evaluations>& poly,
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/multilinear_extension.h:177]:
```suggestion
  static bool WriteTo(const math::MultilinearExtension<Evaluations>& mle,
```
How about like this for consistency?

File: tachyon/math/polynomials/multivariate/multilinear_extension.h [hunk 30]
```diff
@@ -166,7 +168,33 @@ H AbslHashValue(H h, const MultilinearExtension<Evaluations>& mle) {
...
+                      Buffer* buffer) {
+    return buffer->Write(poly.evaluations());
+  }
+
+  static bool ReadFrom(const ReadOnlyBuffer& buffer,
+                       math::MultilinearExtension<Evaluations>* poly) {
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/multilinear_extension.h:183]:
```suggestion
                       math::MultilinearExtension<Evaluations>* mle) {
```
How about like this for consistency?

File: tachyon/math/polynomials/multivariate/multilinear_extension.h [hunk 31]
```diff
@@ -166,7 +168,33 @@ H AbslHashValue(H h, const MultilinearExtension<Evaluations>& mle) {
...
+    Evaluations evaluations;
+    if (!buffer.Read(&evaluations)) return false;
+    *poly = math::MultilinearExtension<Evaluations>(std::move(evaluations));
+    return true;
+  }
+
+  static size_t EstimateSize(
+      const math::MultilinearExtension<Evaluations>& poly) {
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/multilinear_extension.h:191]:
```suggestion
      const math::MultilinearExtension<Evaluations>& mle) {
```
How about like this for consistency?

File: tachyon/math/polynomials/multivariate/linear_combination.h [hunk 32]
```diff
@@ -0,0 +1,125 @@
...
+  // max number of evaluations for each term
+  size_t max_evaluations_ = 0;
+  // number of variables of the polynomial
+  size_t num_variables_ = 0;
+  // list of coefficients and their corresponding list of indexes in reference
+  // to evaluations in |flattened_ml_evaluations_|
+  std::vector<LinearCombinationTerm<F>> terms_;
+  // holds shared pointers of unique multilinear evaluations
+  std::vector<std::shared_ptr<MLE>> flattened_ml_evaluations_;
+  // holds all unique dense ML evaluations currently used + their
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/linear_combination.h:105]:
```suggestion
  // holds all unique dense multi linear evaluations currently used + their
```

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+using PolyPtr = std::shared_ptr<Poly>;
+
+class LinearCombinationTest : public FiniteFieldTest<F> {};
+
+}  // namespace
+
+// Tests |AddTerm()| and |SumOfProducts()|
+TEST_F(LinearCombinationTest, AddTermAndSumOfProducts) {
+  LinearCombination<Poly> linear_combination(kMaxDegree);
+  F coefficient_1 = F::Random();
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:27]:
```suggestion
  F coefficient1 = F::Random();
```

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+
+  // |evals| takes in 5 random Polys and an additional 5 of the same Random
+  // Poly |test_multiple| to test logic of inputting the same random Poly
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:30]:
```suggestion
  // |evals| takes in 5 random polys and an additional 5 of the same random
  // poly |test_multiple| to test logic of inputting the same random poly
```

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 35]
```diff
@@ -0,0 +1,68 @@
...
+  PolyPtr test_multiple = std::make_shared<Poly>(Poly::Random(kMaxDegree));
+  std::vector<PolyPtr> evals_1 =
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:32]:
```suggestion
  std::vector<PolyPtr> evals1 =
```

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 36]
```diff
@@ -0,0 +1,68 @@
...
+      base::CreateVector(10, [test_multiple](size_t i) {
+        if (i % 2 != 0) return std::make_shared<Poly>(Poly::Random(kMaxDegree));
+        return test_multiple;
+      });
+  linear_combination.AddTerm(coefficient_1, evals_1);
+
+  Point evaluation_point = base::CreateVector(kMaxDegree, F::Random());
+
+  // Set up the expected evaluation result for this 1 term
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:41]:
```suggestion
  // Set up the expected evaluation result for this term 1
```

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 37]
```diff
@@ -0,0 +1,68 @@
...
+  F expected_evaluation =
+      coefficient_1 *
+      std::accumulate(evals_1.begin(), evals_1.end(), F::One(),
+                      [&evaluation_point](F& acc, const PolyPtr eval) {
+                        return acc *= eval->Evaluate(evaluation_point);
+                      });
+
+  // Test validity of |AddTerm()| and |SumOfProducts()| on 1 term
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:49]:
```suggestion
  // Test validity of |AddTerm()| and |SumOfProducts()| on term 1
```

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 38]
```diff
@@ -0,0 +1,68 @@
...
+  EXPECT_EQ(linear_combination.SumOfProducts(evaluation_point),
+            expected_evaluation);
+
+  // Test |AddTerm()| and |SumOfProducts| on 2 terms
```
Comment (TomTaehoonKim) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:53]:
```suggestion
  // Test |AddTerm()| and |SumOfProducts| on term 2
```

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc (hunk 36 above)
Comment (ashjeong) [reply tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:41]: i wanted to say it was the case of one term, not specifically this term 1 with coefficient1 and evals1

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc (hunk 37 above)
Comment (ashjeong) [reply tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:49]: ^^

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc (hunk 38 above)
Comment (ashjeong) [reply tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:53]: should be for 2 terms (term 1& 2) not just term 2

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 39]
```diff
@@ -0,0 +1,68 @@
...
+using PolyPtr = std::shared_ptr<Poly>;
+
+class LinearCombinationTest : public FiniteFieldTest<F> {};
+
+}  // namespace
+
+// Tests |AddTerm()| and |SumOfProducts()|
+TEST_F(LinearCombinationTest, AddTermAndSumOfProducts) {
+  LinearCombination<Poly> linear_combination(kMaxDegree);
+  F coefficient1 = F::Random();
```
Comment (Insun35) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:27]: I think `coefficient_1` would be better for consistency(<-> `coefficient_2`).
```suggestion
  F coefficient_1 = F::Random();
```

File: tachyon/math/polynomials/multivariate/linear_combination_unittest.cc [hunk 40]
```diff
@@ -0,0 +1,68 @@
...
+
+  // |evals| takes in 5 random polys and an additional 5 of the same random
+  // poly |test_multiple| to test logic of inputting the same random poly
+  PolyPtr test_multiple = std::make_shared<Poly>(Poly::Random(kMaxDegree));
+  std::vector<PolyPtr> evals1 =
```
Comment (Insun35) [inline tachyon/math/polynomials/multivariate/linear_combination_unittest.cc:32]:
```suggestion
  std::vector<PolyPtr> evals_1 =
```

**Review by Insun35 — CHANGES_REQUESTED** [review]

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM
